### ISSUE TITLE
Feature request: Provide notation for declaring arguments and minimally inserted implicit

### ISSUE THREAD

**Issue by RalfJung** [coq/coq#7767]
#### Version

8.8.0

#### Operating system
Linux

#### Description of the problem

There are several cases where minimally inserted implicit arguments are preferable over maximally inserted ones: With maximal insertion, just writing `term` will desugar to `@term _ _`, which leads to surprising behavior like `eapply ... (arg:=name)` breaking in sections (https://github.com/coq/coq/issues/4536) or not working at all for implicit arguments (https://github.com/coq/coq/issues/6029). If the argument is a typeclass, the hard-to-control order and time of TC resolution can also cause problems with maximally inserted implicit arguments (so in the Iris proof mode, we also sometimes have to write `iApply @lemma` instead of `iApply lemma`).

So, I generally think the first argument of a function, if implicit, should be minimally inserted to avoid such issue. That's a matter of style, of course, and I am not saying Coq's standard library should change -- but Coq currently makes it hard to follow this style. One has to write `Arguments lemma [_] ...` to declare arguments as minimally inserted implicit, so every single lemma would have to come with an `Arguments` declaration.

Hence I am proposing to add a new syntax `[var : type]` to declare an argument and make it minimally inserted implicit, matching the `{var : type}` one can write to declare them maximally inserted implicit. For example, I should be able to write:
```
Definition test [T : Type] (x : T) := x.
Check test. (* Prints test : forall T : Type, T -> T. *)

Section foo.
  Context [T : Type].
  Definition test2 (x : T) := x.
End foo.
Check test2. (* Prints test : forall T : Type, T -> T. *)
```
`` `[ ... ]`` should also work just like `` `{ ... }`` does but make the arguments minimally inserted.

(I am not sure which feature requests need to be CEPs so I figured I'd start with an easier-to-write entry in the issue tracker.)

**Comment by herbelin** [comment]
@SimonBoulier started a [branch](https://github.com/SimonBoulier/coq/tree/print_implicit_args2) provided this wish at the last Coq developer meeting.

**Comment by SimonBoulier** [comment]
I confirm. And the change seems very simple.
@herbelin Do you want that I do this?

**Comment by herbelin** [comment]
@SimonBoulier: Personally, I would find that this would provide a nice unity with `Arguments`. I don't know the opinion of others though.

PS:  Regarding #4536 and #6029 though, the problem is in my opinion elsewhere. It would be more natural that the `with` notation apply to a reference rather than to a term.

**Comment by SimonBoulier** [comment]
WIP here: https://github.com/SimonBoulier/coq/tree/non_maximal_implicit

**Comment by maximedenes** [comment]
I believe this can be closed now that #11235 was merged.
